pallets/allocations/src/lib.rs

```diff
@@ -77,84 +82,99 @@ pub mod pallet {
 	pub struct Pallet<T>(PhantomData<T>);
 
 	#[pallet::hooks]
-	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_runtime_upgrade() -> Weight {
+			remove_storage_prefix(<Pallet<T>>::name().as_bytes(), b"CoinsConsumed", b"");
+			T::DbWeight::get().writes(1)
+		}
+	}
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
-		/// Can only be called by an oracle, trigger a coin creation and an event
-		#[pallet::weight(
-			<T as pallet::Config>::WeightInfo::allocate(proof.len() as u32)
-		)]
-		// we add the `transactional` modifier here in the event that one of the
-		// transfers fail. the code itself should already prevent this but we add
-		// this as an additional guarantee.
+		/// Optimized allocation call, which will batch allocations of various amounts
+		/// and destinations and together. This allow us to be much more efficient and thus
+		/// increase our chain's capacity in handling these transactions.
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::batch(batch.len().try_into().unwrap_or(T::MaxAllocs::get())))]
 		#[transactional]
-		pub fn allocate(
+		pub fn batch(
 			origin: OriginFor<T>,
-			to: T::AccountId,
-			amount: BalanceOf<T>,
-			proof: Vec<u8>,
+			batch: BoundedVec<(T::AccountId, BalanceOf<T>), T::MaxAllocs>,
 		) -> DispatchResultWithPostInfo {
 			Self::ensure_oracle(origin)?;
-			ensure!(
-				!pallet_emergency_shutdown::Pallet::<T>::shutdown(),
-				Error::<T>::UnderShutdown
-			);
-			ensure!(
-				amount >= T::ExistentialDeposit::get().saturating_mul(2u32.into()),
-				Error::<T>::DoesNotSatisfyExistentialDeposit,
-			);
 
-			if amount == Zero::zero() {
-				return Ok(Pays::No.into());
+			// sanity checks
+			let mut full_issuance: BalanceOf<T> = Zero::zero();
+			for (_account, amount) in batch.iter().cloned() {
+				ensure!(
+					amount >= T::ExistentialDeposit::get().saturating_mul(2u32.into()),
```

Comment:
Not related to this PR, but I wonder why we have checked 2 x ED here? We probably don't need to check the `amount` at all because ED would only matter for to the resulted balance and not the amount we are allocating. If we need to have a minimum allocation, it would be better to be like `MaxAllocs` configurable.